File: python/Q435_non_overlapping_intervals.py
#
# @lc app=leetcode id=435 lang=python3
#
# [435] Non-overlapping Intervals
#

# @lc code=start
class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        intervals.sort()

        result: int = 0
        current_end: int = intervals[0][1]

        for i in range(1, len(intervals)):
            start: int = intervals[i][0]
            end: int = intervals[i][1]

            if start >= current_end:
                current_end = end
            else:
                result += 1
                current_end = min(current_end, end)

        return result


# @lc code=end

File: rust/src/leetcode/Q435_non_overlapping_intervals.rs
/*
 * @lc app=leetcode id=435 lang=rust
 *
 * [435] Non-overlapping Intervals
 */

// @lc code=start
impl Solution {
    pub fn erase_overlap_intervals(intervals: Vec<Vec<i32>>) -> i32 {
        let mut intervals: Vec<Vec<i32>> = intervals;
        intervals.sort();

        let mut result: i32 = 0;
        let mut end: i32 = intervals[0][1];

        for interval in intervals.iter().skip(1) {
            if interval[0] < end {
                result += 1;
                end = end.min(interval[1]);
            } else {
                end = interval[1];
            }
        }

        result
    }
}
// @lc code=end
